src/EFCore.PG/Migrations/NpgsqlMigrationsSqlGenerator.cs

```diff
@@ -551,30 +546,40 @@ protected override void Generate(RenameTableOperation operation, [CanBeNull] IMo
                 .Append(" ON ")
                 .Append(Dependencies.SqlGenerationHelper.DelimitIdentifier(operation.Table, operation.Schema));
 
-            if (method != null)
+            if (operation[NpgsqlAnnotationNames.IndexMethod] is string method && method.Length > 0)
             {
                 builder
                     .Append(" USING ")
                     .Append(method);
             }
 
+            var operators = operation[NpgsqlAnnotationNames.IndexOperators] as string[];
+
             builder
                 .Append(" (")
                 .Append(IndexColumnList(operation.Columns, operators))
                 .Append(")");
 
-            if (!string.IsNullOrEmpty(operation.Filter))
```

Comment:
@roji hi,

I don't know why this 'Where' is deleted but i can't add a filter for my all entities. I need to add this index filter 'is_deleted is FALSE' just because i am using soft delete with unique indexes.  Any idea?

Reply:
/CC: @khellang

Reply:
`INCLUDE` isn't the same as the `WHERE` option. The feature which @khellang added supports the first one to improve read performance (the specified columns are included into the index, but not participate in ordering; this allows to take values only from one table called an index table without joining to the original table).

Reply:
@suadev It's deleted because it's using the overriden `IndexOptions` method, with a call to `base.IndexOptions` below.

If you take a look at the base class (in EF Core), you can see that it's adding the `WHERE` clause if there's a filter specified:

https://github.com/dotnet/efcore/blob/c13f8b4a645c6b7cf141c9c62833ea4269622912/src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs#L1832-L1843

Reply:
To add to what @YohDeadfall wrote, the WHERE support wasn't deleted - it was simply moved into EF Core itself (see [MigrationsSqlGenerator](https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Migrations/MigrationsSqlGenerator.cs#L1837).

Are you seeing an actual issue with it not working? If so, please open a new issue with a code sample so we can investigate.

Reply:
Make sense. But i can not add filter to my unique indexes like below. It doesnt work.  How can i do it?

```csharp 
protected override void IndexOptions(CreateIndexOperation operation, IModel model, MigrationCommandListBuilder builder)
        {
            if (operation.IsUnique)
            {
                operation.Filter = "is_deleted is FALSE";
            }
        }
```

Reply:
@suadev Are you trying to swap out the default migrations SQL generator with your own custom version, which automatically adds a `WHERE` clause to all unique indexes?

Reply:
@khellang  exactly :)

Reply:
added  base.Generate(operation, model, builder)  but still doesn't work.
```csharp
       protected override void IndexOptions(CreateIndexOperation operation, IModel model, MigrationCommandListBuilder builder)
        {
            operation.Filter = "is_deleted is FALSE";
            base.Generate(operation, model, builder);
        }
```

Reply:
@suadev Is it being called at all?

Reply:
Probably not but i am not sure. When i run 'dotnet ef migrations add', my custom MigrationSqlGenerator intercept it, right? So i don't know how can i debug it.

Reply:
Can you show the wire-up of your custom generator? Are you doing it like this?

https://github.com/npgsql/efcore.pg/blob/43694356fa07f839ca3503e2718a80240353229f/src/EFCore.PG/Extensions/NpgsqlServiceCollectionExtensions.cs#L75

Reply:
I thought you asked IndexOptions method.  Yes my custom sqlgenerator works, i am sure because we did another manipulations before.  I user 'ReplaceService' like below.

```csharp
            services.AddDbContext<TenantDbContext>((serviceProvider, optionsBuilder) =>
            {
                var tenantProvider = serviceProvider.GetRequiredService<ITenantProvider>();
                var tenantSchema = tenantProvider.GetSchema();

                var connectionSettings = configuration.GetConnectionString();
                optionsBuilder.UseNpgsql(connectionSettings.Default,
                        option =>
                        {
                            option.MigrationsHistoryTable("EFMigrationsHistory", tenantSchema);
                            option.EnableRetryOnFailure(maxRetryCount: 5,
                                maxRetryDelay: TimeSpan.FromSeconds(30),
                                errorCodesToAdd: null);
                        })
                    .ReplaceService<IMigrationsSqlGenerator, PmsMigrationSqlGenerator>()
                    .EnableSensitiveDataLogging(isDevelopmentEnvironment);
            });
```

Reply:
You may need to look at this: https://docs.microsoft.com/en-us/ef/core/miscellaneous/configuring-dbcontext#design-time-dbcontext-configuration

Reply:
@roji thanks, but i didn't get it. I am still not sure what is my mistake.  Overriding IndexOptions method is not a correct way?

Reply:
You could also try to just modify the model itself, instead of swapping components? I.e.

```csharp
protected override void OnModelCreating(ModelBuilder modelBuilder) {
	foreach (var entity in modelBuilder.Model.GetEntityTypes()) {
		foreach (var index in entity.GetDeclaredIndexes()) {
			if (index.IsUnique) {
				index.SetAnnotation(RelationalAnnotationNames.Filter, "is_deleted is FALSE");
			}
		}
	}
}
```

Reply:
@khellang thank you! i am not even aware of GetDeclaredIndexes() method !

Reply:
Changing the model is definitely a better approach than replacing services, wherever it's possible.